cpp_api/objects/equation.cpp

```diff
@@ -2,355 +2,333 @@
 #include "equation.h"
 
 
-EQ* create_equation_deep_copy(EQ* original_equation)
+void Equation::copy_from_EQ_obj(const EQ* obj)
 {
-    EQ* copy_eq = (EQ*) SW_nalloc(sizeof(EQ));
+    this->lec = copy_char_array(obj->lec);
     // NOTE : we do not use memcpy() because memcpy() actually makes  
     //        a shallow copy of a struct instead of a deep copy
-    copy_eq->clec = clec_deep_copy(original_equation->clec);
-    copy_eq->lec = copy_char_array(original_equation->lec);
-    copy_eq->solved = original_equation->solved;
-    copy_eq->method = original_equation->method;
+    this->clec = clec_deep_copy(obj->clec);
```

Comment:
@jmpplan @yvda23  
It is only yesterday (25/10/2023) that I understood that the C function `memcpy()` actually creates a `shallow copy` of a C object, not a deep copy !

If a copied structure contains an attribute which is a pointer, only the adress is copied, not its content. So, after running `memcpy()`, we have two objects with a internal pointer pointing to the same adress in memory.
So, when one the two object is freed, the other contains a pointer pointing to invalid memory adress !

To make a long story short, `memcpy()` is a powerful function but dangerous if you don't know what you are doing.

Reply:
Indeed! When you have a complex object in C, you always have to create a `copy_object()` function (as well as a `free_object()` function). And sometimes recursively ☹ .